### PR TITLE
#854 fix platform installation only offered if port is selected

### PR DESCRIPTION
**Pull request by davegarthsimpson**
### Motivation
Users reported that when selecting a board without a port, the auto-install prompt was erroneously not being shown.

### Change description
Ensures we are showing the prompt when desired, whilst satisfying the aim of [#485](https://github.com/arduino/arduino-ide/pull/485) which sought to prevent an "already closed" auto-install prompt being shown when a board is unplugged. Also includes refactoring to improve readability and logic modularity.

### Reviewer checklist

* [ ] PR addresses a single concern.
* [ ] The PR has no duplicates (please search among the [Pull Requests](https://github.com/arduino/arduino-ide/pulls) before creating one)
* [ ] PR title and description are properly filled.
* [ ] Docs have been added / updated (for bug fixes / features)

### CODE REVIEW

**Review by ubidefeo — COMMENTED** [review]
I don't get the prompt to install the board support on Mac OS 12.3 Monterey

Also Mac having the standard Bluetooth port it gets selected automatically so testing this is not really possible
CleanShot 2022-06-30 at 16 11 16@2x

**Comment by davegarthsimpson** [comment]
> I don't get the prompt to install the board support on Mac OS 12.3 Monterey
> 
> Also Mac having the standard Bluetooth port it gets selected automatically so testing this is not really possible CleanShot 2022-06-30 at 16 11 16@2x

Hi @ubidefeo. In this case you're not seeing the prompt because of another issue [this one](https://github.com/arduino/arduino-ide/issues/979), which I have also resolved but in [another PR](https://github.com/arduino/arduino-ide/pull/1131l). Bear in mind the issues overlap, so testing for [979](https://github.com/arduino/arduino-ide/issues/979) requires a port is selected due to the PR that this issue is related to.

**Review by per1234 — APPROVED** [review]
I have verified that this fixes https://github.com/arduino/arduino-ide/issues/854, while also not resulting in a regression of https://github.com/arduino/arduino-ide/issues/1100

Thanks Dave!

**Review by kittaakos — COMMENTED** [review]
From time to time, it does not work. I uploaded a screencast [here](https://gofile.io/d/VCSbfA).

I have the `avr` platform installed, and when I select `ATmega*` board, the installer does not propose anything.

**Comment by davegarthsimpson** [comment]
> From time to time, it does not work. I uploaded a screencast [here](https://gofile.io/d/VCSbfA).
> 
> I have the `avr` platform installed, and when I select `ATmega*` board, the installer does not propose anything.

Hey @kittaakos this might be for the same issue that I responded to with @ubidefeo above. I'll merge main into this to verify.

**Comment by kittaakos** [comment]
I thought the missing ports were due to this:

```
root ERROR Uncaught Exception:  Error: 14 UNAVAILABLE: Error starting board discoveries: []
root ERROR Error: 14 UNAVAILABLE: Error starting board discoveries: []
    at Object.callErrorFromStatus (/Users/a.kitta/dev/git/arduino-ide/node_modules/@grpc/grpc-js/build/src/call.js:31:26)
    at Object.onReceiveStatus (/Users/a.kitta/dev/git/arduino-ide/node_modules/@grpc/grpc-js/build/src/client.js:409:49)
    at Object.onReceiveStatus (/Users/a.kitta/dev/git/arduino-ide/node_modules/@grpc/grpc-js/build/src/client-interceptors.js:328:181)
    at /Users/a.kitta/dev/git/arduino-ide/node_modules/@grpc/grpc-js/build/src/call-stream.js:187:78
    at processTicksAndRejections (node:internal/process/task_queues:78:11)
```

My issue seemed to be unrelated.

**Comment by davegarthsimpson** [comment]
> I thought the missing ports were due to this:

I was referring to this:

> Hi @ubidefeo. In this case I think you're not seeing the prompt because of another issue [this one](https://github.com/arduino/arduino-ide/issues/979), which I have also resolved but in [another PR](https://github.com/arduino/arduino-ide/pull/1131l). Bear in mind the issues overlap, so testing for resolution of [979](https://github.com/arduino/arduino-ide/issues/979) requires a port is selected due to the PR that this issue is related to.

**Comment by kittaakos** [comment]
> From time to time, it does not work.

It looks OK now. Thank you!

File: arduino-ide-extension/src/browser/boards/boards-service-provider.ts [hunk 1]
```diff
@@ -209,7 +220,7 @@ export class BoardsServiceProvider implements FrontendApplicationContribution {
     }
   }
 
-  protected async tryReconnect(): Promise<boolean> {
+  protected tryReconnect(): boolean {
```
Comment (davegarthsimpson) [inline arduino-ide-extension/src/browser/boards/boards-service-provider.ts:223]: couldn't see a reason for this to be async, it's not called across the wire either

**Review by kittaakos — APPROVED** [review]
I tried it, and it worked. Thank you!

**Pull request merged**